flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/over/NonTimeUnboundedPrecedingFunction.java

```diff
@@ -0,0 +1,591 @@
+                out.collect(output);
+                iterator.remove();
+                valueMapState.remove(curKey.getLong(keyIdx));
+                currFunction.retract(curValue);
+                isRetracted = true;
+            }
+        }
+
+        // Update sorted key state without the retracted row
+        sortedKeyState.update(sortedKeyList);
+    }
+
+    private void removeFromSortedListWithValueMapAndAccumulatorMap(
+            RowData input, Collector<RowData> out) throws Exception {
+        boolean isRetracted = false;
+        input.setRowKind(RowKind.INSERT);
+        List<RowData> sortedKeyList = sortedKeyState.value();
+        ListIterator<RowData> iterator = sortedKeyList.listIterator();
+
+        while (iterator.hasNext()) {
+            RowData curKey = iterator.next();
+            RowData curValue = valueMapState.get(curKey.getLong(keyIdx));
+            if (isRetracted) {
+                RowData prevAcc = accMapState.get(curKey.getLong(keyIdx));
+                if (prevAcc.equals(currFunction.getValue())) {
+                    // If previously accumulated value is same as the newly accumulated value
+                    // then no need to send updates downstream since nothing has changed
+                    break;
+                }
+                // Emit updated agg value for all records after retraction
+                output.setRowKind(RowKind.UPDATE_BEFORE);
+                output.replace(curValue, prevAcc);
+                out.collect(output);
+
+                currFunction.accumulate(curValue);
+                accMapState.put(curKey.getLong(keyIdx), currFunction.getValue());
+                output.setRowKind(RowKind.UPDATE_AFTER);
+                output.replace(curValue, currFunction.getValue());
+                out.collect(output);
+            } else if (equaliser.equals(curValue, input)) {
+                // Retract record
+                RowData curAcc = accMapState.get(curKey.getLong(keyIdx));
+                currFunction.setAccumulators(curAcc);
+                output.setRowKind(RowKind.UPDATE_BEFORE);
+                output.replace(input, currFunction.getValue());
+                out.collect(output);
+                iterator.remove();
+
+                valueMapState.remove(curKey.getLong(keyIdx));
+                accMapState.remove(curKey.getLong(keyIdx));
+                if (iterator.hasPrevious()) {
+                    RowData prev = sortedKeyList.get(iterator.previousIndex());
+                    RowData prevAcc = accMapState.get(prev.getLong(keyIdx));
+                    currFunction.setAccumulators(prevAcc);
+                } else {
+                    currFunction.setAccumulators(currFunction.createAccumulators());
+                }
+                isRetracted = true;
+            }
+        }
+
+        // Update sorted key state without the retracted row
+        sortedKeyState.update(sortedKeyList);
```

Comment:
this could be moved up to the statements that update the `valueMapState` and `accMapState`.
That would keep all related state updates together.

Reply:
Moved to `insertRow` helper function